opm/simulators/linalg/FlexibleSolver.hpp

```diff
@@ -89,33 +100,41 @@ class FlexibleSolver : public Dune::InverseOperator<VectorTypeT, VectorTypeT>
 
     // Machinery for making sequential or parallel operators/preconditioners/scalar products.
     template <class Comm>
-    void initOpPrecSp(const MatrixType& matrix, const boost::property_tree::ptree& prm, const Comm& comm)
+    void initOpPrecSp(const MatrixType& matrix, const boost::property_tree::ptree& prm,
+                      const std::function<VectorTypeT()> weightsCalculator, const Comm& comm)
     {
         // Parallel case.
         using ParOperatorType = Dune::OverlappingSchwarzOperator<MatrixType, VectorType, VectorType, Comm>;
+        using pt = const boost::property_tree::ptree;
         auto linop = std::make_shared<ParOperatorType>(matrix, comm);
         linearoperator_ = linop;
+        auto child = prm.get_child_optional("preconditioner");
         preconditioner_
-            = Opm::PreconditionerFactory<ParOperatorType, Comm>::create(*linop, prm.get_child("preconditioner"), comm);
+            = Opm::PreconditionerFactory<ParOperatorType, Comm>::create(*linop, child? *child : pt(),
+                                                                        weightsCalculator, comm);
         scalarproduct_ = Dune::createScalarProduct<VectorType, Comm>(comm, linearoperator_->category());
     }
 
-    void initOpPrecSp(const MatrixType& matrix, const boost::property_tree::ptree& prm, const Dune::Amg::SequentialInformation&)
+    void initOpPrecSp(const MatrixType& matrix, const boost::property_tree::ptree& prm,
+                      const std::function<VectorTypeT()> weightsCalculator, const Dune::Amg::SequentialInformation&)
     {
         // Sequential case.
         using SeqOperatorType = Dune::MatrixAdapter<MatrixType, VectorType, VectorType>;
+        using pt = const boost::property_tree::ptree;
         auto linop = std::make_shared<SeqOperatorType>(matrix);
         linearoperator_ = linop;
-        preconditioner_ = Opm::PreconditionerFactory<SeqOperatorType>::create(*linop, prm.get_child("preconditioner"));
+        auto child = prm.get_child_optional("preconditioner");
+        preconditioner_ = Opm::PreconditionerFactory<SeqOperatorType>::create(*linop, child? *child : pt(),
+                                                                              weightsCalculator);
         scalarproduct_ = std::make_shared<Dune::SeqScalarProduct<VectorType>>();
     }
 
     void initSolver(const boost::property_tree::ptree& prm)
     {
-        const double tol = prm.get<double>("tol");
-        const int maxiter = prm.get<int>("maxiter");
-        const int verbosity = prm.get<int>("verbosity");
-        const std::string solver_type = prm.get<std::string>("solver");
+        const double tol = prm.get<double>("tol", 1e-2);
```

Comment:
I am a bit doubtful about the defaulting here. I think we should make life easy for users, but if we allow defaulting we must somehow print the actual options applied (default or not) in the log. When we add printing the parameter tree to the log, that would miss out on these defaults.

If this is motivated by the weird behaviour (timestep chop death) currently caused by errors in the json file, I think that we should rather be cautious here and not allow defaults, but fix that weirdness to print proper errors.

Reply:
I do not agree at all. These json files are so complicated that even the authors of the FlexibleSolvers sometimes do not phrase them correctly. The options are bound to change (e.g. new ones added). For the user writing a self contained json file is an immense error. Let us face it, most of the options that we have will never be changed (and probably should not). Not having defaults simply makes it rather unusable.

It took me a lot of willpower to continue working on this when I experienced the time step chopping. If I would have been a nornal user I would just have abandoned it and as a new user maybe OPM, too. 

When we print the parameter files, we simply set up the defaults at the same time and warn for unused options. But that is for another PR and hopefully different developer. We should not trade usability for preventing developer errors...